### PR TITLE
Feature/sale discount code

### PR DESCRIPTION
**Pull request by sunnylochun**

### CODE REVIEW

**Comment by sunnylochun** [comment]
[(2) 【銷售管理】訂單顯示多顯示折扣代碼於折價券名稱之後](https://trello.com/c/gyC3cEBj/2781-2-%E3%80%90%E9%8A%B7%E5%94%AE%E7%AE%A1%E7%90%86%E3%80%91%E8%A8%82%E5%96%AE%E9%A1%AF%E7%A4%BA%E5%A4%9A%E9%A1%AF%E7%A4%BA%E6%8A%98%E6%89%A3%E4%BB%A3%E7%A2%BC%E6%96%BC%E6%8A%98%E5%83%B9%E5%88%B8%E5%90%8D%E7%A8%B1%E4%B9%8B%E5%BE%8C)

File: src/components/sale/SaleCollectionAdminCard.tsx [hunk 1]
```diff
@@ -341,4 +348,68 @@ const SaleCollectionAdminCard: React.VFC<{
   )
 }
 
+const DiscountCode: React.VFC<{ type: string; target: string }> = ({ type, target }) => {
+  const apolloClient = useApolloClient()
+  const [code, setCode] = useState('')
+
+  const getDiscountCode = async ({ type, target }: { type: string; target: string }) => {
```
Comment (futianshen) [inline src/components/sale/SaleCollectionAdminCard.tsx:355]: can set `Promise<string>` as `ReturnType` of function

File: src/components/sale/SaleCollectionAdminCard.tsx [hunk 2]
```diff
@@ -341,4 +348,68 @@ const SaleCollectionAdminCard: React.VFC<{
...
+          fetchPolicy: 'no-cache',
+        })
+        return voucher.voucher_by_pk?.voucher_code.code || ''
+      default:
+        return ''
+    }
+  }
+
+  useEffect(() => {
+    getDiscountCode({ type, target }).then(value => value && setCode(value))
```
Comment (futianshen) [inline src/components/sale/SaleCollectionAdminCard.tsx:387]: NITS: can just use `setCode` as the callback function

```ts
getDiscountCode({ type, target }).then(setCode)
```

File: src/components/sale/SaleCollectionAdminCard.tsx [hunk 3]
```diff
@@ -341,4 +348,68 @@ const SaleCollectionAdminCard: React.VFC<{
...
+          hasura.GET_COUPON_CODE_BY_COUPON,
+          hasura.GET_COUPON_CODE_BY_COUPONVariables
+        >({
+          query: GET_COUPON_CODE_BY_COUPON,
+          variables: {
+            id: target,
+          },
+          fetchPolicy: 'no-cache',
+        })
+        return coupon.coupon_by_pk?.coupon_code.code || ''
```
Comment (futianshen) [inline src/components/sale/SaleCollectionAdminCard.tsx:368]: it should throw an error here if the code is empty

File: src/components/sale/SaleCollectionAdminCard.tsx [hunk 4]
```diff
@@ -341,4 +348,68 @@ const SaleCollectionAdminCard: React.VFC<{
...
+          hasura.GET_VOUCHER_CODE_BY_VOUCHER,
+          hasura.GET_VOUCHER_CODE_BY_VOUCHERVariables
+        >({
+          query: GET_VOUCHER_CODE_BY_VOUCHER,
+          variables: {
+            id: target,
+          },
+          fetchPolicy: 'no-cache',
+        })
+        return voucher.voucher_by_pk?.voucher_code.code || ''
```
Comment (futianshen) [inline src/components/sale/SaleCollectionAdminCard.tsx:380]: same as above

File: src/components/sale/SaleCollectionAdminCard.tsx [hunk 5]
```diff
@@ -341,4 +348,68 @@ const SaleCollectionAdminCard: React.VFC<{
...
+      default:
+        return ''
```
Comment (futianshen) [inline src/components/sale/SaleCollectionAdminCard.tsx:382]: same as above

Comment (sunnylochun) [reply src/components/sale/SaleCollectionAdminCard.tsx:382]: discount type will not only be voucher or coupon
other type like card and downPrice maybe just return empty value

File: src/components/sale/SaleCollectionAdminCard.tsx (hunk 3 above)
Comment (sunnylochun) [reply src/components/sale/SaleCollectionAdminCard.tsx:368]: throw error or 
return empty message something like 無折扣碼 ？

Comment (futianshen) [reply src/components/sale/SaleCollectionAdminCard.tsx:368]: To reduce unnecessary requests, I think we should only send requests when permission is allowed (ex: only app-owner has permission here)
if we get nothing from the response then throw an error

File: src/components/sale/SaleCollectionAdminCard.tsx (hunk 5 above)
Comment (futianshen) [reply src/components/sale/SaleCollectionAdminCard.tsx:382]: only coupons and vouchers have a discount code, so I think it shouldn't be handled in `getDiscountCode`

**Pull request merged**